File: pages/refactorings/replace-mutable-fields-with-lenses/02-rename-tree/Before.scala
// Uppercase the variable of every Var node: a recursive walk that
// rebuilds each hit by hand.
object Before:
  case class Var(name: String, ref: Int)
  enum Expr:
    case EVar(v: Var)
    case EApp(f: Expr, x: Expr)
    case ELam(bind: String, body: Expr)

  import Expr.*

  def renameAll(e: Expr): Expr = e match
    case EVar(v)     => EVar(v.copy(name = v.name.toUpperCase))
    case EApp(f, x)  => EApp(renameAll(f), renameAll(x))
    case ELam(b, bd) => ELam(b, renameAll(bd))

File: pages/refactorings/replace-mutable-fields-with-lenses/02-rename-tree/Before.hs
-- Uppercase the variable of every Var node: a recursive walk that
-- rebuilds each hit by hand.
module Before where

import Data.Char (toUpper)

data Var  = Var  { vName :: String, vRef :: Int }
  deriving (Eq, Show)
data Expr = EVar Var | EApp Expr Expr | ELam String Expr
  deriving (Eq, Show)

renameAll :: Expr -> Expr
renameAll (EVar v)     = EVar v { vName = map toUpper (vName v) }
renameAll (EApp f x)   = EApp (renameAll f) (renameAll x)
renameAll (ELam b bd)  = ELam b (renameAll bd)
